py/desispec/io/emlinefit.py

```diff
@@ -0,0 +1,567 @@
+#!/usr/bin/env python
+
+"""
+desispec.io.emlinefit
+====================
+IO routines for desi_emlinefit_afterburner.
+"""
+
+import os
+import sys
+from astropy.io import fits
+import fitsio
+import numpy as np
+from scipy.optimize import curve_fit
+from desitarget.geomask import match_to
+from desiutil.dust import ext_odonnell, ext_fitzpatrick
+from desiutil.dust import ebv as dust_ebv
+from desiutil.log import get_logger
+from matplotlib.backends.backend_pdf import PdfPages
+import matplotlib.pyplot as plt
+from matplotlib import gridspec
+
+allowed_emnames = ["OII", "HDELTA", "HGAMMA", "HBETA", "OIII", "HALPHA"]
+
+def get_targetids(redrockfn, bitnames, log=get_logger()):
```

Comment:
This function opens and reads the redrock file twice to perform a selection of targetids.  I think it would be cleaner to make this a non-I/O function that takes a fibermap and bitnames and returns targetids (or maybe the indices of the fibermap or a boolean array of matches).

Some of this logic looks like it would be cleaner to use `desitarget.targets.main_cmx_or_sv` to figure out what the relevant target column is and what mask to import.  Please check if that would work.

Reply:
Here, I've kept the function as is (for the time being).

Maybe that doesn't really matter as the `redrock*fits` files are small, but by my experience, I think it's faster to proceed this way:
- first only read the file structure with `fits.open()`, but not the data, to inspect the column content (that step is super-quick)
- then read only the relevant columns.
than just reading the whole data.
Same for `desitarget.targets.main_cmx_or_sv`: that function assumes that the whole data have been read.

Note that in the new code version, I also use that two-steps approach, e.g.:
https://github.com/desihub/desispec/blob/febddeed3ab1885b56b956c5ada69dd2526b6b8c/py/desispec/io/emlinefit.py#L148-L164

Please let me know if there is a better way to proceed.